File: src/bumperbot_utils/bumperbot_utils/safety_stop.py
#!/usr/bin/env python3
from enum import Enum
import rclpy
from rclpy.node import Node
from sensor_msgs.msg import LaserScan
from std_msgs.msg import Bool

class State(Enum):
    FREE = 0
    WARNING = 1
    DANGER = 2

class SafetyStop(Node):
    def __init__(self):
        super().__init__("safety_stop_node")

        self.declare_parameter("danger_distance", 0.2)
        self.declare_parameter("scan_topic", "scan")
        self.declare_parameter("safety_stop_topic", "safety_stop")

        self.danger_distance = self.get_parameter("danger_distance").get_parameter_value().double_value
        self.scan_topic = self.get_parameter("scan_topic").get_parameter_value().string_value
        self.safety_stop_topic = self.get_parameter("safety_stop_topic").get_parameter_value().string_value

        self.state = State.FREE

        self.laser_sub = self.create_subscription(LaserScan, self.scan_topic, self.laser_callback, 10)
        self.safety_stop_pub = self.create_publisher(Bool, self.safety_stop_topic, 10)

    def laser_callback(self, msg: LaserScan):
        self.state = State.FREE

        for range_value in msg.ranges:
            if range_value <= self.danger_distance:
                self.state = State.DANGER
                break

        is_safety_stop = Bool()

        if self.state == State.DANGER:
            is_safety_stop.data = True
        elif self.state == State.FREE:
            is_safety_stop.data = False

        self.safety_stop_pub.publish(is_safety_stop)

def main():
    rclpy.init()
    node = SafetyStop()
    rclpy.spin(node)
    rclpy.shutdown()

if __name__ == "__main__":
    main()

File: src/bumperbot_utils/src/safety_stop.cpp
#include "rclcpp/rclcpp.hpp"
#include "sensor_msgs/msg/laser_scan.hpp"
#include "std_msgs/msg/bool.hpp"

using std::placeholders::_1;

enum State 
{
    FREE = 0,
    WARNING, 
    DANGER
};

class SafetyStop : public rclcpp::Node 
{
    public:
        SafetyStop() : Node("safety_stop_node")
        {
            declare_parameter<double>("danger_distance", 0.2);
            declare_parameter<std::string>("scan_topic", "scan");
            declare_parameter<std::string>("safety_stop_topic", "safety_stop");

            danger_distance_ = get_parameter("danger_distance").as_double();

            std::string scan_topic = get_parameter("scan_topic").as_string();
            std::string safety_stop_topic  = get_parameter("safety_stop_topic").as_string();

            state_ = State::FREE;

            laser_sub_ = create_subscription<sensor_msgs::msg::LaserScan>(scan_topic, 10, std::bind(&SafetyStop::laserCallback, this, _1));

            safety_stop_pub_ = create_publisher<std_msgs::msg::Bool>(safety_stop_topic, 10);
        }

    private:
        double danger_distance_;
        State state_;
        rclcpp::Subscription<sensor_msgs::msg::LaserScan>::SharedPtr laser_sub_;
        rclcpp::Publisher<std_msgs::msg::Bool>::SharedPtr safety_stop_pub_;

        void laserCallback(const sensor_msgs::msg::LaserScan & msg)
        {
            state_ = State::FREE;

            for (const auto & range : msg.ranges)
            {
                if(range <= danger_distance_)
                {
                    state_ = State::DANGER;
                    break;
                }
            }

            std_msgs::msg::Bool is_safety_stop;
            if(state_ == State::DANGER) 
            {
                is_safety_stop.data = true;
            } else if(state_ == State::FREE)
            {
                is_safety_stop.data = false;
            }

            safety_stop_pub_->publish(is_safety_stop);
        }
};

int main(int argc, char *argv[])
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<SafetyStop>());
    rclcpp::shutdown();

    return 0;
}